### PR TITLE
Binary accuracy bug fixes - Metric accuracy method refactor

### PR DESCRIPTION
**Pull request by dfossl**
Related to PR #16010 

This fixes BinaryAccuracy metric class to function with sample weights and refactors metrics.py and metrics_utils.py such that metrics_utils.py has new methods with behaviour that better reflects what [MeanMetricWrapper.update_state()](https://github.com/keras-team/keras/blob/master/keras/metrics/base_metric.py#L603) is expecting. The Accuracy classes in metrics now use these methods instead. The accuracy methods now all wrap the metrics_utils.py methods. Binary accuracy is the only one that has an additional process, this allows it to behave as expected from the public-facing side. Having this separation allows for public-facing accuracy functions to be edited for public-facing behaviour without affecting the Metric Accuracy classes directly.

## Binary Accuracy fix:
As described previously in #16010 
Binary accuracy also had an issue with sample_weight when you pass batches of sequences.

If you give binary accuracy a batches of sequences it functions fine.
```python
y_true = [[1, 1, 0, 0],
         [1, 1, 0, 0]]
y_pred = [[1, 1, 0, 0],
         [1, 1, 0, 1]]
m = tf.keras.metrics.BinaryAccuracy()
m.reset_state()
m.update_state(y_true, y_pred)
print(m.result().numpy())
```
Returns
```python
0.875
```
But applying weights throws an error
```python
y_true = [[1, 1, 0, 0],
         [1, 1, 0, 0]]
y_pred = [[1, 1, 0, 0],
         [1, 1, 0, 1]]
sample_weight = [[1, 1, 0, 0],
                  [1, 1, 0, 0]]
m = tf.keras.metrics.BinaryAccuracy()
m.reset_state()
m.update_state(y_true, y_pred, sample_weight)
print(m.result().numpy())
```
```python
InvalidArgumentError: {{function_node __wrapped__Squeeze_device_/job:localhost/replica:0/task:0/device:CPU:0}} Can not squeeze dim[1], expected a dimension of 1, got 4 [Op:Squeeze]
```
Turns out this is because binary_accuracy is calling backend.mean
```python
return backend.mean(tf.equal(y_true, y_pred), axis=-1)
```
The issue is this is being called before sample_weight gets applied. Simply casting to an int tensor allows for sample_weight to be handled appropriately.
```python
return tf.cast(tf.equal(y_true, y_pred), backend.floatx())
```
This results in expected behaviour:

```python
y_true = [[1, 1, 0, 0],
         [1, 1, 0, 0]]
y_pred = [[1, 1, 0, 0],
         [1, 1, 0, 1]]
sample_weight = [[1, 1, 0, 0],
                  [1, 1, 0, 0]]
m = tf.keras.metrics.BinaryAccuracy()
m.reset_state()
m.update_state(y_true, y_pred, sample_weight)
print(m.result().numpy())
```
Returns
```python
1.0
```

binary_accuracy() in metrics.py still calls backend.mean() but BinaryAccuracy() class now uses metrics_utils method binary_matches which returns the casted tensor.

### CODE REVIEW

File: keras/metrics/metrics.py [hunk 1]
```diff
@@ -3194,6 +3194,10 @@ def accuracy(y_true, y_pred):
 def binary_accuracy(y_true, y_pred, threshold=0.5):
   """Calculates how often predictions match binary labels.
 
+  Note: calls metrics_utils.binary_matches with mean reduction. This maintains
+  public facing binary_accuracy behavior and seperates it from the vital
+  behavior of the binary_matches method needed in backend dependencies.
```
Comment (fchollet) [inline keras/metrics/metrics.py:3199]: This note is related to internal implementation details and should not appear in the docstring, which is the public-facing documentation of the function.

File: keras/metrics/metrics.py [hunk 2]
```diff
@@ -3211,17 +3215,19 @@ def binary_accuracy(y_true, y_pred, threshold=0.5):
   Returns:
     Binary accuracy values. shape = `[batch_size, d0, .. dN-1]`
   """
-  y_pred = tf.convert_to_tensor(y_pred)
-  threshold = tf.cast(threshold, y_pred.dtype)
-  y_pred = tf.cast(y_pred > threshold, y_pred.dtype)
-  return backend.mean(tf.equal(y_true, y_pred), axis=-1)
+
+  return backend.mean(metrics_utils.binary_matches(y_true, y_pred, threshold), axis=-1)
```
Comment (fchollet) [inline keras/metrics/metrics.py:3219]: Note: you can actually use `tf.reduce_mean` instead of `backend.mean`

File: keras/metrics/metrics.py [hunk 3]
```diff
@@ -3211,17 +3215,19 @@ def binary_accuracy(y_true, y_pred, threshold=0.5):
...
 
 
 @keras_export('keras.metrics.categorical_accuracy')
 @tf.__internal__.dispatch.add_dispatch_support
 def categorical_accuracy(y_true, y_pred):
   """Calculates how often predictions match one-hot labels.
 
+  Note: wraps metrics_utils.categorical_matches. This seperates public facing
```
Comment (fchollet) [inline keras/metrics/metrics.py:3227]: Same comment applies to all similar notes in the PR

File: keras/utils/metrics_utils.py [hunk 4]
```diff
@@ -839,3 +839,123 @@ def _assert_splits_match(nested_splits_lists):
       for splits_list in nested_splits_lists[1:]
       for (s1, s2) in zip(nested_splits_lists[0], splits_list)
   ]
+
+
+def binary_matches(y_true, y_pred, threshold=0.5):
+  """Creates int Tensor, 1 for label-prediction match, 0 for mismatch.
+
+  Args:
+    y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`.
```
Comment (fchollet) [inline keras/utils/metrics_utils.py:848]: ", of shape `(batch_size, d0, ... dN)`"

File: keras/utils/metrics_utils.py [hunk 5]
```diff
@@ -839,3 +839,123 @@ def _assert_splits_match(nested_splits_lists):
...
+    y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`.
```
Comment (fchollet) [inline keras/utils/metrics_utils.py:849]: Same

File: keras/utils/metrics_utils.py [hunk 6]
```diff
@@ -839,3 +839,123 @@ def _assert_splits_match(nested_splits_lists):
...
+  y_pred = tf.math.argmax(y_pred, axis=-1)
+
+  # If the predicted output and actual output types don't match, force cast them
+  # to match.
+  if backend.dtype(y_pred) != backend.dtype(y_true):
+    y_pred = tf.cast(y_pred, backend.dtype(y_true))
+  return tf.cast(tf.equal(y_true, y_pred), backend.floatx())
+
+
+def categorical_matches(y_true, y_pred):
```
Comment (fchollet) [inline keras/utils/metrics_utils.py:892]: I am not sure this utility is needed, you could just inline its contents where it is used.

File: keras/utils/metrics_utils.py [hunk 7]
```diff
@@ -839,3 +839,123 @@ def _assert_splits_match(nested_splits_lists):
...
+      dtype=backend.floatx())
+
+  # returned matches is expected to have same shape as y_true input
+  if reshape_matches:
+    return tf.reshape(matches, shape=y_true_org_shape)
+
+  return matches
+
+
+def top_k_categorical_matches(y_true, y_pred, k=5):
```
Comment (fchollet) [inline keras/utils/metrics_utils.py:948]: I am not sure this utility is needed, you could just inline its contents where it is used.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR -- this is a valuable refactor.

**Comment by fchollet** [comment]
Since this PR involves fixing some behaviors, please also add a unit test highlighting the fix (it should be a test that would have failed before this PR).

File: keras/metrics/metrics.py (hunk 1 above)
Comment (dfossl) [reply keras/metrics/metrics.py:3199]: Noted will fix, my bad.

File: keras/utils/metrics_utils.py (hunk 6 above)
Comment (dfossl) [reply keras/utils/metrics_utils.py:892]: Just checking to make sure I understand.
Do you mean to delete this utility method and where categorical_matches is called just call sparse_categorical_matches but argmax the y_true when passing it into the sparse method? Sorry if I'm misinterpreting.

Comment (fchollet) [reply keras/utils/metrics_utils.py:892]: Yes, exactly.

**Comment by fchollet** [comment]
Please let us know when the PR is ready for review again.

**Comment by dfossl** [comment]
> Please let us know when the PR is ready for review again.

Good to go. There were no failing tests before. But I added tests for the new util methods. I tried my best to model the tests after similar ones for the accuracy methods in metrics.py.  Let me know if there are any bad practices I fell into and I'll fix them.👍

Also if there is something I am supposed to do to switch from Draft to PR let me know. First time I have made a PR into Draft so not too familiar with it.

Cheers.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the nice contribution!

**Pull request merged**